cleanrl/cql_continuous/sac_cql.py

```diff
@@ -0,0 +1,396 @@
+    parser.add_argument('--lagrange-thresh', type=float, default=0.0,
+                        help='Lagrange threshold for automatic tuning of the alpha_prime component')
+    parser.add_argument('--with-lagrange', type=lambda x:bool(strtobool(x)), default=False, nargs='?', const=True,
+                        help='automatic tuning of the alpha_prime coefficient (min-q-weight) coefficient.')
+    
+    args = parser.parse_args()
+    if not args.seed:
+        args.seed = int(time.time())
+
+    # create offline gym id: 'BeamRiderNoFrameskip-v4' -> 'beam-rider-expert-v0'
+    args.offline_gym_id = re.sub(r'(?<!^)(?=[A-Z])', '-', args.gym_id).lower().replace(
+        "v2", "") + args.offline_dataset_id
+
+# TRY NOT TO MODIFY: setup the environment
+experiment_name = f"{args.gym_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
+writer = SummaryWriter(f"runs/{experiment_name}")
+writer.add_text('hyperparameters', "|param|value|\n|-|-|\n%s" % (
+        '\n'.join([f"|{key}|{value}|" for key, value in vars(args).items()])))
+if args.prod_mode:
+    import wandb
+    wandb.init(project=args.wandb_project_name, entity=args.wandb_entity, sync_tensorboard=True, config=vars(args), name=experiment_name, monitor_gym=True, save_code=True)
+    writer = SummaryWriter(f"/tmp/{experiment_name}")
+
+# TRY NOT TO MODIFY: seeding
+device = torch.device('cuda' if torch.cuda.is_available() and args.cuda else 'cpu')
+env = gym.make(args.gym_id)
+random.seed(args.seed)
+np.random.seed(args.seed)
+torch.manual_seed(args.seed)
+torch.backends.cudnn.deterministic = args.torch_deterministic
+env.seed(args.seed)
+env.action_space.seed(args.seed)
+env.observation_space.seed(args.seed)
+input_shape = env.observation_space.shape[0]
+output_shape = env.action_space.shape[0]
+# respect the default timelimit
+assert isinstance(env.action_space, Box), "only continuous action space is supported"
+if args.capture_video:
+    env = Monitor(env, f'videos/{experiment_name}')
+
+# ALGO LOGIC: initialize agent here:
+LOG_STD_MAX = 2
+LOG_STD_MIN = -5
+
+def layer_init(layer, weight_gain=1, bias_const=0):
+    if isinstance(layer, nn.Linear):
+        if args.weights_init == "xavier":
+            torch.nn.init.xavier_uniform_(layer.weight, gain=weight_gain)
+        elif args.weights_init == "orthogonal":
+            torch.nn.init.orthogonal_(layer.weight, gain=weight_gain)
+        if args.bias_init == "zeros":
+            torch.nn.init.constant_(layer.bias, bias_const)
+
+class Policy(nn.Module):
+    def __init__(self, input_shape, output_shape, env):
+        super(Policy, self).__init__()
+        self.fc1 = nn.Linear(input_shape, 256) # Better result with slightly wider networks.
+        self.fc2 = nn.Linear(256, 256)
+        self.mean = nn.Linear(256, output_shape)
+        self.logstd = nn.Linear(256, output_shape)
+        # action rescaling
+        self.register_buffer("action_scale", 
```

Comment:
So what if we replace `action_scale` with just python scalar, would it be ok? Becasue `action_scale` is not really being updated, so you could do things like

`action_scale = (env.action_space.high - env.action_space.low) / 2.)`

Reply:
I now recalled the logic behind having the action scale and the action bias as registered buffer:

First, when trying `action_scale = (env.action_space.high - env.action_space.low) / 2.)`, there is the error of multiplying a tensor by the scalar `self.action_scale`, as per the following error log:
```
python sac_cql.py 
Warning: Flow failed to import. Set the environment variable D4RL_SUPPRESS_IMPORT_ERROR=1 to suppress this message.
No module named 'flow'
Warning: CARLA failed to import. Set the environment variable D4RL_SUPPRESS_IMPORT_ERROR=1 to suppress this message.
No module named 'carla'
pybullet build time: Nov 26 2019 09:35:14
/home/d055/anaconda3/envs/cleanrl/lib/python3.6/site-packages/gym/logger.py:30: UserWarning: WARN: Box bound precision lowered by casting to float32
  warnings.warn(colorize('%s: %s'%('WARN', msg % args), 'yellow'))
/home/d055/anaconda3/envs/cleanrl/lib/python3.6/site-packages/gym/logger.py:30: UserWarning: WARN: Box bound precision lowered by casting to float32
  warnings.warn(colorize('%s: %s'%('WARN', msg % args), 'yellow'))
Traceback (most recent call last):
  File "sac_cql.py", line 281, in <module>
    next_state_actions, next_state_log_pi, _ = pg.get_action(s_next_obses)
  File "sac_cql.py", line 174, in get_action
    action = y_t * self.action_scale + self.action_bias
  File "/home/d055/anaconda3/envs/cleanrl/lib/python3.6/site-packages/torch/tensor.py", line 480, in __array__
    return self.numpy()
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first
```

I believe this is what was addressed by having `self.action_scale` and `self.action_bias` as th.FloatTensor as in:
https://github.com/vwxyzjn/cleanrl/blob/c1fb51a7c32596ef3133af4617f92d2095051c66/cleanrl/sac_continuous_action.py#L139

However, this seems to require an overload of the `to(device, ...)` method, so that `self.action_scale` and `self.action_bias` are on the same device as the `y_t` used in `action = y_t * self.action_scale + self.action_bias`:
https://github.com/vwxyzjn/cleanrl/blob/c1fb51a7c32596ef3133af4617f92d2095051c66/cleanrl/sac_continuous_action.py#L171

The idea behind using `register_buffer` was thus to remove the need for having to declare a FloatTensor and moving it to device via overload of the `to()` method.
**Namely, any registered buffer will be automatically copied to the `device` specified when directly calling the `to()` method of the mother class.**, which I find to be a useful trick as further simplification of the code.

Reply:
From additional experiments, using register_buffer properly helps minimize calls to the `.to(device)` function, making the code run slightly faster.